src/accountsdb/buffer_pool.zig

```diff
@@ -774,91 +968,260 @@ pub const CachedRead = struct {
         }
     };
 
-    pub fn readByte(self: CachedRead, index: usize) u8 {
-        std.debug.assert(self.frame_indices.len != 0);
-        std.debug.assert(index < self.len());
-        const offset = index + self.first_frame_start_offset;
-        std.debug.assert(offset >= self.first_frame_start_offset);
+    pub fn readByte(self: ReadHandle, index: usize) u8 {
+        return switch (self.inner) {
+            .cached => |cached| blk: {
+                std.debug.assert(cached.frame_indices.len != 0);
+                std.debug.assert(index < self.len());
+                const offset = index + cached.first_frame_start_offset;
+                std.debug.assert(offset >= cached.first_frame_start_offset);
+                std.debug.assert(offset / FRAME_SIZE < cached.frame_indices.len);
+                std.debug.assert(
+                    cached.frame_indices[offset / FRAME_SIZE] < cached.buffer_pool.frames.len,
+                );
+
+                break :blk cached.buffer_pool.frames[
+                    cached.frame_indices[offset / FRAME_SIZE]
+                ][offset % FRAME_SIZE];
+            },
+            .sub_read => |sub_read| sub_read.parent.readByte(index + sub_read.start),
+            .external => |external| external.slice[index],
+        };
+    }
+
+    /// External to the BufferPool
+    pub fn initExternal(data: []u8) ReadHandle {
+        return ReadHandle{
+            .inner = .{ .external = .{ .slice = data, .owned = false } },
+        };
+    }
+
+    /// External to the BufferPool, data will be freed upon .deinit
+    pub fn initExternalOwned(data: []u8) ReadHandle {
+        return ReadHandle{
+            .inner = .{ .external = .{ .slice = data, .owned = true } },
+        };
+    }
 
-        return self.buffer_pool.frames[
-            self.frame_indices[offset / FRAME_SIZE]
-        ][offset % FRAME_SIZE];
+    fn initCached(
+        buffer_pool: *BufferPool,
+        frame_indices: []const FrameIndex,
+        first_frame_start_offset: FrameOffset,
+        last_frame_end_offset: FrameOffset,
+    ) ReadHandle {
+        return .{
+            .inner = .{
+                .cached = .{
+                    .buffer_pool = buffer_pool,
+                    .frame_indices = frame_indices,
+                    .first_frame_start_offset = first_frame_start_offset,
+                    .last_frame_end_offset = last_frame_end_offset,
+                },
+            },
+        };
+    }
+
+    pub fn dupeExternalOwned(self: ReadHandle, allocator: std.mem.Allocator) !ReadHandle {
```

Comment:
do we really need to handle this as a readHandle -- feels like once they want to get an owned Account, it can just be an Account (unrelated to the ReadHandle) -- we could remove the External enum from all this logic